### DISCUSSION TITLE
support for screen scripts

### DISCUSSION THREAD

**Discussion by wforumw** [ltsp/ltsp#205] [Q&A]
Hi,
In the old LTSP5 we could use screen scripts (SCREEN_xx) each accessible by pressing the Ctrl-Alt-Fx
We used it to start remote desktop sessions.
Is this still possible or is there a better way to connect some clients to Windows Servers.?
Many thanks

**Comment by alkisg** [comment]
Hello,

the initial thought is that the new LTSP won't support screen scripts. In the distant past, it was easy; /etc/inittab handled VTs, and anything not defined there was free for X to use; that got more complicated later on, and in the end, LTSP5 had race conditions of e.g. Xorg not starting until the user manually switched to vt7 etc.

Different VTs are normally used by one user (not multiple users at the same time), to log in to Xorg or to xfreerdp or to SSH etc. AFAIK session files and workspaces can be used to get more or less the same result. For example, the sysadmin may provide a /usr/share/xsessions/kiosk.desktop or xfreerdp.desktop session, and users may select them from the display manager when they need them. The ltsp.conf AUTOLOGIN parameter and a dummy temporary user may also be used whenever appropriate.

That said, if someone is sure that different VTs are better for their use case, I don't object to including code for this upstream in LTSP, but we'll need to find a way that works with systemd, wayland and all the display managers; we want the "normal" way to do it that would work without LTSP as well, not the old LTSP5 way that ignored everything and had race conditions.

**Comment by wforumw** [comment]
Thanks! We used it before so users could choose if the use Linux or Windows (via rdesktop)

**Comment by alkisg** [comment]
One method is to tell them "login to Linux anyway, and if you want to use Windows, run this launcher/menu that runs xfreerdp full screen with sound/usb forwarding etc etc".

Another idea is to create a /usr/share/xsessions/xfreerdp.desktop session, and have a special "windows" user, and when users select it at the display manager login screen (it can be passwordless), the "windows" user will login, and again run xfreerdp full screen etc.

And a third idea is to try to launch another xorg at a different VT.
Which one of those ideas sounds more suitable to you?

**Comment by FaberfoX** [comment]
I'll chime in with my use case, as I'm in the process of upgrading from a quite old install.
I'm using a screen script for vt7 to directly connect to a windows terminal server, auto login-in with a username based on the ip address of the station, so, no sessions, no display manager, just some logic to extract the digits from the hostname and a straight call to xfreerdp. This assumed that X had already started.
What do you think would be the most appropriate way of doing it now? A systemd unit overriding a tty that starts X?

**Comment by alkisg** [comment]
@FaberfoX, one way to do this is with something like the following in ltsp.conf:

```shell
# Create a user that is local to the client, so that authentication/SSHFS/NFS aren't needed:
PRE_INIT_KIOSK_USER="mkdir -p /run/home && useradd --base-dir=/run/home --comment='kiosk user' --create-home --shell=/bin/bash --system kiosk"

# Enable autologin for that user:
AUTOLOGIN=kiosk
RELOGIN=kiosk

# Create an appropriate session in /etc/ltsp, then symlink it:
POST_INIT_XFREERDP_SESSION="ln -s /etc/ltsp/xfreerdp.desktop /usr/share/xsessions/xfreerdp.desktop"

# And set it as the default session of this user:
POST_INIT_KIOSK_DMRC='printf "[Desktop]\nSession=xfreerdp\n" > /run/home/kiosk/.dmrc'
```

So with these (and possibly a couple more that I forgot), lightdm will start, it won't display a login screen, but it will properly launch a session and register it with systemd-logind etc, and autologin the user, and local devices will properly work and things won't run as root etc.

I propose that you start writing and maintaining a wiki page about this in the community wiki:
https://github.com/ltsp/community/wiki/xfreerdp

And that you document all the steps, exactly as they fit you, and I'll help resolve any outstanding issues, and the rest of the LTSP users will benefit from your wiki page.

**Comment by FaberfoX** [comment]
@alkisg, working on it. I had a minimal install in a chroot, had to install openbox as otherwise lightdm was trying to pull over 400 packages and wasn't starting if I did a no-install-recommends. Now lightdm is autostarting but it's complaining that "User ltsp can't auth via pamltsp", can you point me on what to edit?
Once I get this working I'll do the wiki page.

**Comment by alkisg** [comment]
In the POST_INIT_XFREERDP_SESSION parameter in my previous message, I had a quoting mistake (used double quotes inside double quotes). I fixed it and I tested all of this and it worked for me.

I used this /etc/ltsp/xfreerdp.desktop:

```shell
[Desktop Entry]
Name=XTERM
Comment=This session just runs an XTERM
Exec=xterm
TryExec=xterm
Icon=xterm
Type=Application
DesktopNames=XTERM
Keywords=launch;XTERM;desktop;session;
```

This doesn't run xfreerdp. It just runs xterm (install it if you don't have it). Then from that xterm, you can play with xfreerdp and its parameters (e.g. xfreerdp /v:server /f /u:''), and once you get this right, just modify the Exec= line in /etc/ltsp/xfreerdp.desktop, and run `ltsp initrd` to propagate it to the client.

P.S. after you get this going, you may optimize it by running the lighter "nodm" instead of "lightdm"; I can add code in LTSP that makes AUTOLOGIN work in nodm too at that point.

**Comment by solune** [comment]
I have this desktop file:
```
[Desktop Entry]
Name=VMWAREVIEW
Comment=This session just runs vmware-view
Exec=/etc/ltsp/viewonly
TryExec=/etc/ltsp/viewonly
Icon=vmware-view
Type=Application
DesktopNames=VMWAREVIEW
Keywords=launch;VMWAREVIEW;desktop;session;
```

and this is my viewonly script:
```
x-window-manager &
        xset s off         # don't activate screensaver
        xset -dpms         # disable DPMS (Energy Star) features.
        xset s noblank     # don't blank the video device
        exec /usr/bin/vmware-view
/sbin/init 0
```
As you can see, I want to shutdown the computer if the user close the vmware-view application.
How can I do this ?

**Comment by alkisg** [comment]
```shell
x-window-manager &
        xset s off         # don't activate screensaver
        xset -dpms         # disable DPMS (Energy Star) features.
        xset s noblank     # don't blank the video device
        /usr/bin/vmware-view
/sbin/poweroff
```

**Comment by solune** [comment]
It works !
Thanks a lot

**Comment by alkisg** [comment]
I think there's enough information in the wiki and in issues currently for LTSP users to be able to implement their own screen sessions. Moving this issue to community.

**Comment by jphein** [comment]
@alkisg Has this changed at all since you added SESSION_SCRIPT?
https://github.com/ltsp/ltsp/issues/18#issuecomment-550353273

I'm having trouble with the session showing up in the login menu. Let alone calling it automatically.
I'm using Ubuntu Bionic and Ubuntu 19.10.

**Comment by alkisg** [comment]
@jphein, I haven't committed SESSION_SCRIPT yet, so no changes there.

Open a terminal on the client and check your /usr/share/xsession scripts, if the session's there, it should show up in the DM screen.

**Comment by jphein** [comment]
It just autoboots a normal session for me. The session file is there now.

```
kiosk@ltsp103:/usr/share/xsessions$ ls
kodi.desktop  plasma.desktop  ubuntu.desktop  xfreerdp.desktop
kiosk@ltsp103:/usr/share/xsessions$ cat xfreerdp.desktop 
[Desktop Entry]
Name=XTERM
Comment=This session just runs an XTERM
Exec=xterm
TryExec=xterm
Icon=xterm
Type=Application
DesktopNames=XTERM
Keywords=launch;XTERM;desktop;session;
kiosk@ltsp103:/usr/share/xsessions$ 
```

I'll turn of autoboot and manually try to call the session file.

**Comment by jphein** [comment]
Manually selecting the XTERM session worked. 
So just the part of selecting the default session didn't work. 
Do we set that somewhere else in the ltsp.conf?

I created the https://github.com/ltsp/community/wiki/xfreerdp entry.

**Comment by jphein** [comment]
cat /etc/X11/default-display-manager
/usr/sbin/gdm3

**Comment by jphein** [comment]
`sudo apt install lightdm`
Switching to lightdm still did not not auto start the XTERM session. 
Oops, I didn't run `sudo ltsp image /`
I also had to add `LIGHTDM_CONF='user-session=xfreerdp'`
to [clients] part of ltsp.conf and run `sudo ltsp initrd`
It does work using LightDM. 

However, GDM3 works much better for me. Do you have an example for GDM3? Maybe I can use ltsp.conf to choose which display manager to use? Use GDM on bare metal server, and LightDM or NoDM for clients?

**Comment by alkisg** [comment]
Linking to a similar question/answer: https://github.com/ltsp/ltsp/discussions/514#discussioncomment-1258238

**Comment by alkisg** [comment]
Hi @svetoi1985, currently the LTSP discussions are English-only.
Please click "edit" in your message and use google translate to write it in English.

In the future, if someone wants to support non-English languages, they can ask me to create a new category for them; e.g. I might create a "Greek support" category later on. But the "General", "QA" etc categories will always be English-only.

**Reply by svetoi1985** [reply]
ok
